mp/src/game/shared/momentum/run/mom_run_safeguards.cpp

```diff
@@ -0,0 +1,198 @@
+#include "cbase.h"
+
+#include "mom_run_safeguards.h"
+#include "in_buttons.h"
+#include "mom_player_shared.h"
+
+#ifdef CLIENT_DLL
+#include "input.h"
+#endif
+
+#include "tier0/memdbgon.h"
+
+static MAKE_CONVAR(mom_run_safeguard_doublepress_maxtime, "0.5", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Controls the max amount of time (in seconds) that a double press is registered for run safeguards "
+    "with the doublepress safeguard mode active.",
+    0.0f, 5.0f);
+
+static MAKE_CONVAR(mom_run_safeguard_practicemode, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for enabling practice mode during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_NONE, RUN_SAFEGUARD_MODE_DOUBLEPRESS);
+
+static MAKE_CONVAR(mom_run_safeguard_restart, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for restarting the map during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_NONE, RUN_SAFEGUARD_MODE_DOUBLEPRESS);
+
+static MAKE_CONVAR(mom_run_safeguard_saveloc_tele, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for teleporting to a saved location during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_NONE, RUN_SAFEGUARD_MODE_DOUBLEPRESS);
+
+static MAKE_CONVAR(mom_run_safeguard_chat_open, "1", FCVAR_ARCHIVE | FCVAR_REPLICATED,
+    "Changes the safeguard for opening chat during a run.\n"
+    "0 = OFF,\n"
+    "1 = Enable when not pressing any movement keys,\n"
+    "2 - Enable on double press.\n",
+    RUN_SAFEGUARD_MODE_NONE, RUN_SAFEGUARD_MODE_DOUBLEPRESS);
+
+CRunSafeguard::CRunSafeguard(const char *szAction) : m_flLastTimePressed(0.0f), m_bDoublePressSafeguard(true)
+{
+    Q_strncpy(m_szAction, szAction, sizeof(m_szAction));
+}
+
+bool CRunSafeguard::IsSafeguarded(RunSafeguardMode_t mode)
+{
+    CMomentumPlayer *pPlayer;
+#ifdef CLIENT_DLL
+    pPlayer = C_MomentumPlayer::GetLocalMomPlayer();
+#elif GAME_DLL
+    pPlayer = CMomentumPlayer::GetLocalPlayer();
+#endif
+    if (!pPlayer)
+        return false;
+
+    CMomRunEntityData *pEntData;
+#ifdef CLIENT_DLL
+    pEntData = pPlayer->GetCurrentUIEntData();
+#elif GAME_DLL
+    pEntData = pPlayer->GetRunEntData();
+#endif
```

Comment:
It's better to do `else` in this case because it's either client or server, plus this would catch the case where the pointer wasn't properly initialized for the nullcheck:
```suggestion
    CMomRunEntityData *pEntData
#ifdef CLIENT_DLL
        = pPlayer->GetCurrentUIEntData();
#else
        = pPlayer->GetRunEntData();
#endif

```